Socialanalytics

Kama93 opened the pull request:
Hi Daniel,
I created a new file, socialanalytics.clj, in the directory conexp.fca.applications. Right now, this file provides the following functionality:
For a formal context, there are function to compute the average shortest path lengths for the following graphs:
- The one having the objects and attributes as verticies and the incidents as edges.
- The one having the objects as verticies and in which two objects share an edge if they share an attribute.
- The one having the attributes as verticies and in which two attributes share an edge if they share an object.

I therefore used the algrorithmm of Floyd-Warshall. As this algorithm works on the adjacency-matricies of the graphs, I implemented functions to compute the adjacency-matricies of the above discribed graphs.

For the three given types of graphs I also implemented functions to compute the vertice-degrees.
Therefore, I implemented functions which compute the graphs and return them as adjacency-maps.

Regards,
Max

exot commented:
Cool! :)

I merged your commits into a separate feature branch again (I had to rebase your commits onto `dev` for this).  I will have a look at your code as soon as possible and will let you know about my suggestions, if there will be any.

Best,

  Daniel

Kama93 commented:
Hi Daniel,
I fixed some spelling mistakes:
The singular should always be ``vertex``  instead of ``vertice`` and the plural is ``vertices`` instead of ``verticies``.

Regards,
Max

exot commented:
Hey Max,

sorry for my late reply, I didn't had time to look at your code earlier in detail.

I would propose a minor refactoring:
- Some function names are rather long, and I think there are shorter names that are also telling.  For example, instead of `adjacency-matrix-for-object-and-attribute-projection`, I would propose `combined-projection-adjaceny-matrix`, and likewise `combined-projection`, `combined-projection-shortest-paths`, and `combined-projection-vertex-degrees`.  The same holds true for the other projections, although the name change there would be mostly for consistency.
- Using vectors of vectors to represent matrices is okay, although I would prefer to use native Java arrays for this (even if you need to store the full matrix then).  If you want to keep the current data model, please use transient vectors for speed in conjunction with `assoc!`.
- I would propose to rewrite the computation of average shortest path into a generic `distance-matrix` function that just computes the complete matrix of shortest distances without deleting any rows or columns.  Based on this function, you can then add a `average-shortest-path` function that takes as input a matrix and returns the average shortest path length.
- `object-projection`/`attribute-projection` and `adjacency-matrix-for-attribute-projection`/`adjacency-matrix-for-object-projection` are very similar and should be implemented in terms of a single, more general function (alas, this should be done with a lot of functions in conexp-clj, maybe one day I find the time to do it …).  Once this is done, the function `add-edges` should not be necessary anymore.

In case you have any questions about these proposals, please don't hesitate to ask :)

Looking forward to your changes!

  Daniel

Kama93 commented:
Done=)

Kama93 commented:
Hey Daniel,
after having a little talk to Tom today I think that the name combined-projection is not appropriate. So I renamed combined-projection to context-graph. The names object-projection and attribute-projection are still the same as they are indeed projections, from the graph having the objects and attributes  as nodes (now just named the context-graph) to the one having only the objects or only the attributes as nodes.
Regards,
Max

Kama93 commented:
Hey Daniel,
I implemented some new functionallity: I created methods to compute the k-cores for graphs and contexts. I also implemented an alternative algorithm to compute the average shortest path lengths. It uses breadth-first-search and is therefore very efficient for graphs with a low denisity.

I must say that the work on the k-cores and the new implemantation of avg-path-lengths is based on things that are implemented in the commits of this pull request so it´s not really sensefull to create another pull request now with the new work. My first idea was to wait, until the current PR is closed and its content is merged into dev and open a pull request with the new stuff then.

But as it`s also some content for Social Analytics, I think it would fit quite nice into the socialanalytics branch. 

Do you want met to put the new things in a new PR or should I just push them into the socialanalytics-branch that this pull request is based on, so that you can merge everything together into dev?

Let me know which option you prefer!

Regards,
Max

exot commented:
Hey Max,

first, thanks again for your efforts, all this sounds very promising! :)

Concerning the names of the functions, the only things I am interested in is that they are not too long and they are appropriate.  Thus if `context-graph` is better than `combined-projection`, I certainly won't object ;)

For the additional functionality, I would propose that you branch from your current feature branch for socialanalytics and put the changes there.  As soon as this PR is done, you then can rebase your changes on the new version of `dev`.  The rational behind this approach is that I certainly will have something to complain about (I always have, just ask Tom) and will do some changes to this PR before merging it into `dev`.  If you thus now push more commits into this PR, it will only take longer for me to finish it, which would be a pity.

My current plan is to find some time next week to look at your changes, and then finally merge it into `dev`.

Best,

  Daniel

Kama93 commented:
Hey Daniel,

I am fine with your changes!

Regards,
Max

exot commented:
Then it's done, your changes have now been merged into `dev`.

Thanks again for your work and for your patience! :)

Best,

  Daniel
